File: 03/00_recursao/codes/prefix_parser/prefix_parser.c
#include <stdio.h>

char a[100]; int i;

int eval()
{
  int x = 0;
  while(a[i] == ' ') i++;
  if(a[i] == '+') { i++; return(eval() + eval()); }
  if(a[i] == '*') { i++; return(eval() * eval()); }
  while((a[i] >= '0') && (a[i] <= '9'))
    x = 10*x + (a[i++]-'0');
  return(x);
}

int main(){
  fgets(a,99,stdin);

  printf("%d\n",eval());
}

File: 02/codes/recursao/prefix_parser.go
// Parser (e avaliador) recursivo de expressões em notação prefixa.
// Exemplo de entrada: "+ 3 4" avalia para 7; "* 2 + 3 4" avalia para 14.

package main

import (
	"bufio"
	"fmt"
	"os"
)

var expr string
var pos int

func eval() int {
	x := 0

	for pos < len(expr) && expr[pos] == ' ' {
		pos++
	}

	if pos < len(expr) && expr[pos] == '+' {
		pos++
		return eval() + eval()
	}
	if pos < len(expr) && expr[pos] == '*' {
		pos++
		return eval() * eval()
	}
	for pos < len(expr) && expr[pos] >= '0' && expr[pos] <= '9' {
		x = 10*x + int(expr[pos]-'0')
		pos++
	}
	return x
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	linha, _ := reader.ReadString('\n')
	expr = linha
	pos = 0

	fmt.Println(eval())
}
